bridges/snowbridge/primitives/outbound-queue/src/v2/message.rs

```diff
@@ -0,0 +1,333 @@
+				target: Address::from(target.as_fixed_bytes()),
+				data: Bytes::from(data.to_vec()),
+				value: U256::try_from(*value).unwrap(),
+			}
+			.abi_encode(),
+		}
+	}
+}
+
+/// Representation of a call to the initializer of an implementation contract.
+/// The initializer has the following ABI signature: `initialize(bytes)`.
+#[derive(Clone, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct Initializer {
+	/// ABI-encoded params of type `bytes` to pass to the initializer
+	pub params: Vec<u8>,
+	/// The initializer is allowed to consume this much gas at most.
+	pub maximum_required_gas: u64,
+}
+
+pub trait SendMessage {
+	type Ticket: Clone + Encode + Decode;
+
+	type Balance: BaseArithmetic + Unsigned + Copy;
+
+	/// Validate an outbound message and return a tuple:
+	/// 1. Ticket for submitting the message
+	/// 2. Delivery fee
+	fn validate(message: &Message) -> Result<(Self::Ticket, Self::Balance), SendError>;
+
+	/// Submit the message ticket for eventual delivery to Ethereum
+	fn deliver(ticket: Self::Ticket) -> Result<H256, SendError>;
+}
+
+pub trait GasMeter {
+	/// Measures the maximum amount of gas a command payload will require to *dispatch*, NOT
+	/// including validation & verification.
+	fn maximum_dispatch_gas_used_at_most(command: &Command) -> u64;
+}
+
+/// A meter that assigns a constant amount of gas for the execution of a command
+///
+/// The gas figures are extracted from this report:
+/// > forge test --match-path test/Gateway.t.sol --gas-report
+///
+/// A healthy buffer is added on top of these figures to account for:
+/// * The EIP-150 63/64 rule
+/// * Future EVM upgrades that may increase gas cost
+pub struct ConstantGasMeter;
+
+impl GasMeter for ConstantGasMeter {
+	fn maximum_dispatch_gas_used_at_most(command: &Command) -> u64 {
+		match command {
+			Command::CreateAgent { .. } => 275_000,
+			Command::SetOperatingMode { .. } => 40_000,
+			Command::Upgrade { initializer, .. } => {
+				let initializer_max_gas = match *initializer {
+					Some(Initializer { maximum_required_gas, .. }) => maximum_required_gas,
+					None => 0,
+				};
+				// total maximum gas must also include the gas used for updating the proxy before
+				// the the initializer is called.
+				50_000 + initializer_max_gas
+			},
+			Command::UnlockNativeToken { .. } => 100_000,
+			Command::RegisterForeignToken { .. } => 1_200_000,
+			Command::MintForeignToken { .. } => 100_000,
+			Command::CallContract { gas_limit, .. } => *gas_limit,
+		}
+	}
+}
+
+impl GasMeter for () {
+	fn maximum_dispatch_gas_used_at_most(_: &Command) -> u64 {
+		1
+	}
+}
+
+// Origin for high-priority governance commands
+pub fn primary_governance_origin() -> H256 {
+	hex!("0000000000000000000000000000000000000000000000000000000000000001").into()
+}
+
+// Origin for lower-priority governance commands
+pub fn second_governance_origin() -> H256 {
+	hex!("0000000000000000000000000000000000000000000000000000000000000002").into()
+}
```

Comment:
@yrong these don't really have a place in V2.

Reply:
Remove `second_governance_origin` in https://github.com/vgeddes/polkadot-sdk/pull/6/commits/cc8b6b12119ecb433ef2e0211d34edee718d7eb8

For `primary_governance_origin`, I'll remove it when adding upgrade extrinsic in a separate PR.

Reply:
For V2, `message.origin` should be `AgentIDFor(BridgeHubPara)`, not `primary_governance_origin()`.

`primary_governance_origin()` is really a V1 concept, and needs to to be removed. Its actually an ChannelID, not an AgentId

Reply:
> message.origin should be AgentIDFor(BridgeHubPara)

I'd assume it's `Here` for governance origin.  Some fix in
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/fe9e7043f5feb70a8e81eb8ab5d2cc2d1dc4d704